compiler/backend_inkwell/src/lib.rs

```diff
@@ -11,14 +16,58 @@ use inkwell::{
     AddressSpace,
 };
 
+use candy_frontend::error::{CompilerError, CompilerErrorPayload};
+use candy_frontend::hir;
+use candy_frontend::rich_ir::{RichIr, ToRichIr};
+use candy_frontend::string_to_rcst::ModuleError;
 pub use inkwell;
 use itertools::Itertools;
+use rustc_hash::FxHashSet;
 use std::{
     collections::{HashMap, HashSet},
     path::PathBuf,
     sync::Arc,
 };
 
+#[salsa::query_group(LlvmIrStorage)]
+pub trait LlvmIrDb: OptimizeMir {
+    #[salsa::transparent]
+    fn llvm_ir(&self, module: CandyModule) -> Result<RichIr, ModuleError>;
+}
+
+#[allow(clippy::needless_pass_by_value)]
+fn llvm_ir(db: &dyn LlvmIrDb, module: CandyModule) -> Result<RichIr, ModuleError> {
+    //debug!("{module}: Compiling.");
+    let (mir, errors) = db
+        .optimized_mir(module.clone(), TracingConfig::off())
+        .map(|(mir, _, errors)| (mir, errors))
+        .unwrap_or_else(|error| {
+            let payload = CompilerErrorPayload::Module(error);
+            let mir = Mir::build(|body| {
+                let reason = body.push_text(payload.to_string());
+                let responsible = body.push_hir_id(hir::Id::user());
+                body.push_panic(reason, responsible);
+            });
+            let errors =
+                FxHashSet::from_iter([CompilerError::for_whole_module(module.clone(), payload)]);
+            (Arc::new(mir), Arc::new(errors))
+        });
+
+    if !errors.is_empty() {
+        for error in errors.iter() {
+            println!("{:?}", error);
+        }
+        std::process::exit(1);
+    }
```

Comment:
I think you can just do this:

```suggestion
    let (mir, errors) = db.optimized_mir(module.clone(), TracingConfig::off())?;
```